File: app/src/main/java/com/kelsos/mbrc/services/LibraryServiceImpl.java
package com.kelsos.mbrc.services;

import com.fasterxml.jackson.core.type.TypeReference;
import com.kelsos.mbrc.constants.Protocol;
import com.kelsos.mbrc.data.Page;
import com.kelsos.mbrc.data.PageRange;
import com.kelsos.mbrc.data.library.Album;
import com.kelsos.mbrc.data.library.Artist;
import com.kelsos.mbrc.data.library.Genre;
import com.kelsos.mbrc.data.library.Track;

import java.io.IOException;

import rx.Observable;

public class LibraryServiceImpl extends ServiceBase implements LibraryService {

  @Override
  public Observable<Page<Genre>> getGenres(int offset, int limit) {
    PageRange range = getPageRange(offset, limit);

    return request(Protocol.LibraryBrowseGenres, range == null ? "" : range).flatMap(socketMessage -> Observable.create(subscriber -> {
      try {
        TypeReference<Page<Genre>> typeReference = new TypeReference<Page<Genre>>() {
        };
        Page<Genre> page = mapper.readValue((String) socketMessage.getData(), typeReference);
        subscriber.onNext(page);
        subscriber.onCompleted();
      } catch (IOException e) {
        subscriber.onError(e);
      }
    }));
  }

  @Override
  public Observable<Page<Artist>> getArtists(int offset, int limit) {
    PageRange range = getPageRange(offset, limit);

    return request(Protocol.LibraryBrowseArtists, range == null ? "" : range).flatMap(socketMessage -> Observable.create(subscriber -> {
      try {
        TypeReference<Page<Artist>> typeReference = new TypeReference<Page<Artist>>() {
        };
        Page<Artist> page = mapper.readValue((String) socketMessage.getData(), typeReference);
        subscriber.onNext(page);
        subscriber.onCompleted();
      } catch (IOException e) {
        subscriber.onError(e);
      }
    }));
  }

  @Override
  public Observable<Page<Album>> getAlbums(int offset, int limit) {
    PageRange range = getPageRange(offset, limit);

    return request(Protocol.LibraryBrowseAlbums, range == null ? "" : range).flatMap(socketMessage -> Observable.create(subscriber -> {
      try {
        TypeReference<Page<Album>> typeReference = new TypeReference<Page<Album>>() {
        };
        Page<Album> page = mapper.readValue((String) socketMessage.getData(), typeReference);
        subscriber.onNext(page);
        subscriber.onCompleted();
      } catch (IOException e) {
        subscriber.onError(e);
      }
    }));
  }

  @Override
  public Observable<Page<Track>> getTracks(int offset, int limit) {
    PageRange range = getPageRange(offset, limit);

    return request(Protocol.LibraryBrowseTracks, range == null ? "" : range).flatMap(socketMessage -> Observable.create(subscriber -> {
      try {
        TypeReference<Page<Track>> typeReference = new TypeReference<Page<Track>>() {
        };
        Page<Track> page = mapper.readValue((String) socketMessage.getData(), typeReference);
        subscriber.onNext(page);
        subscriber.onCompleted();
      } catch (IOException e) {
        subscriber.onError(e);
      }
    }));
  }

}

File: app/src/main/kotlin/com/kelsos/mbrc/services/LibraryServiceImpl.kt
package com.kelsos.mbrc.services

import com.fasterxml.jackson.databind.ObjectMapper
import com.kelsos.mbrc.constants.Protocol
import com.kelsos.mbrc.data.Page
import com.kelsos.mbrc.data.library.Album
import com.kelsos.mbrc.data.library.Artist
import com.kelsos.mbrc.data.library.Genre
import com.kelsos.mbrc.data.library.Track
import com.kelsos.mbrc.extensions.toPage
import com.kelsos.mbrc.repository.ConnectionRepository
import com.kelsos.mbrc.utilities.SettingsManager
import io.reactivex.Observable
import javax.inject.Inject

class LibraryServiceImpl
@Inject
constructor(
    repository: ConnectionRepository,
    private val mapper: ObjectMapper,
    settingsManager: SettingsManager
) : ServiceBase(repository, mapper, settingsManager), LibraryService {

  override fun getGenres(offset: Int, limit: Int): Observable<Page<Genre>> {
    val range = getPageRange(offset, limit)
    return request(Protocol.LibraryBrowseGenres, range).flatMap { it.toPage<Genre>(mapper) }
  }

  override fun getArtists(offset: Int, limit: Int): Observable<Page<Artist>> {
    val range = getPageRange(offset, limit)
    return request(Protocol.LibraryBrowseArtists, range).flatMap { it.toPage<Artist>(mapper) }
  }

  override fun getAlbums(offset: Int, limit: Int): Observable<Page<Album>> {
    val range = getPageRange(offset, limit)
    return request(Protocol.LibraryBrowseAlbums, range).flatMap { it.toPage<Album>(mapper) }
  }

  override fun getTracks(offset: Int, limit: Int): Observable<Page<Track>> {
    val range = getPageRange(offset, limit)
    return request(Protocol.LibraryBrowseTracks, range).flatMap { it.toPage<Track>(mapper) }
  }
}
